exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListProofStructureValidator.java

```diff
@@ -181,6 +189,18 @@ private boolean invalidBranch(NodeInfo branch) {
         .allMatch(nodeType -> (nodeType == NodeType.HASHNODE) || (nodeType == NodeType.NONE));
   }
 
+  /**
+   * Returns true if proof of absence node is not a root node in a proof tree.
+   *
+   * @param branches collection of branches info
+   */
+  private boolean proofOfAbsenceIsNotARootNode(List<NodeInfo> branches) {
+    return branches.stream()
```

Comment:
Since we already have depth available during the tree walk,
why don't we accumulate all the absence nodes (as we do with other)
in `#visit(ListProofOfAbsence)` and then in _O(1)_:
(a) verify there is at most one;
(b) if there is one, verify there are no other nodes in the tree;
(c) if there is one, verify its depth is zero.

Also, it seems reasonable to replace separate collections with Multimap<NodeType, NodeInfo>
backed by an enummap, but that can be done separately (an issue?):
```java
    Multimaps.newListMultimap(new EnumMap<NodeType, Collection<NodeInfo>>(NodeType.class),
        ArrayList::new);
```

Reply:
Yes, I thought about making it O(1) instead of O(n) by just checking in `visit(ListProofBranch branchNode)` if either of branch children node types are of absence types but then that check would be done in `visit` and not in `check`. Can do it the way you proposed as well

Reply:
Did that with a boolean field instead.

Reply:
The flag could work, but, as you've said, all the other checks are in a separate method and not in the `visit(...)` methods. It seems reasonable to keep it that way.